From c043b24802cc7420f3af1c72e8fec75b317703cf Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:10:10 -0400
Subject: [usb] xhci: Use correct SLOT ID when handling a reset device command

Message-id: <1410887463-20674-539-git-send-email-dzickus@redhat.com>
Patchwork-id: 95674
O-Subject: [RHEL7 PATCH 538/591] xhci: Use correct SLOT ID when handling a reset device command
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 6fcfb0d682a8212d321a6131adc94daf0905992a
Author: Mathias Nyman <mathias.nyman@linux.intel.com>
Date:   Tue Jun 24 17:14:40 2014 +0300

    xhci: Use correct SLOT ID when handling a reset device command

    Command completion events normally include command completion status,
    SLOT_ID, and a pointer to the original command. Reset device command
    completion SLOT_ID may be zero according to xhci specs 4.6.11.

    VIA controllers set the SLOT_ID to zero, triggering a WARN_ON in the
    command completion handler.

    Use the SLOT ID found from the original command instead.

    This patch should be applied to stable kernels since 3.13 that contain
    the commit 20e7acb13ff48fbc884d5918c3697c27de63922a
    "xhci: use completion event's slot id rather than dig it out of command"

    Cc: stable@vger.kernel.org # 3.13
    Reported-by: Saran Neti <sarannmr@gmail.com>
    Tested-by: Saran Neti <sarannmr@gmail.com>
    Signed-off-by: Mathias Nyman <mathias.nyman@linux.intel.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/host/xhci-ring.c b/drivers/usb/host/xhci-ring.c
index d67ff71..71657d3 100644
--- a/drivers/usb/host/xhci-ring.c
+++ b/drivers/usb/host/xhci-ring.c
@@ -1433,8 +1433,11 @@ static void handle_cmd_completion(struct xhci_hcd *xhci,
   xhci_handle_cmd_reset_ep(xhci, slot_id, cmd_trb, cmd_comp_code);
   break;
  case TRB_RESET_DEV:
-  WARN_ON(slot_id != TRB_TO_SLOT_ID(
-    le32_to_cpu(cmd_trb->generic.field[3])));
+  /* SLOT_ID field in reset device cmd completion event TRB is 0.
+   * Use the SLOT_ID from the command TRB instead (xhci 4.6.11)
+   */
+  slot_id = TRB_TO_SLOT_ID(
+    le32_to_cpu(cmd_trb->generic.field[3]));
   xhci_handle_cmd_reset_dev(xhci, slot_id, event);
   break;
  case TRB_NEC_GET_FW:
-- 
1.7.1